### PR TITLE
Split render compute

### PR DESCRIPTION
**Pull request by Kangz**
PTAL, I'm not entirely happy with how I handled the "common" part in the OpenGL backend, but don't really see a better way. This is also missing the D3D12 changes.

### CODE REVIEW

**Comment by Kangz** [comment]
For #61

File: src/backend/metal/CommandBufferMTL.mm [hunk 1]
```diff
@@ -135,7 +136,8 @@ void EndSubpass() {
 
     void CommandBuffer::FillCommands(id<MTLCommandBuffer> commandBuffer) {
         Command type;
-        Pipeline* lastPipeline = nullptr;
+        PipelineBase* lastPipeline = nullptr;
```
Comment (kainino0x) [inline src/backend/metal/CommandBufferMTL.mm]: is it possible to get rid of lastPipeline and have only lastRenderPipeline and lastComputePipeline?

File: src/backend/RenderPipeline.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+namespace backend {
+
+    // RenderPipelineBase
+
+    RenderPipelineBase::RenderPipelineBase(RenderPipelineBuilder* builder)
+        : PipelineBase(builder), renderPass(std::move(builder->renderPass)), subpass(builder->subpass),
+          inputState(std::move(builder->inputState)), depthStencilState(std::move(builder->depthStencilState)) {
+
+        if (GetStageMask() != (nxt::ShaderStageBit::Vertex | nxt::ShaderStageBit::Fragment)) {
+            builder->HandleError("Graphics pipeline should have exactly a vertex and fragment stage");
```
Comment (austinEng) [inline src/backend/RenderPipeline.cpp]: nit: "Render pipeline" or maybe just "Pipeline"

File: src/backend/d3d12/PipelineD3D12.cpp [hunk 3]
```diff
@@ -162,7 +162,7 @@ namespace d3d12 {
             descriptor.RTVFormats[0] = DXGI_FORMAT_R8G8B8A8_UNORM;
             descriptor.SampleDesc.Count = 1;
 
-            ASSERT_SUCCESS(device->GetD3D12Device()->CreateGraphicsPipelineState(&descriptor, IID_PPV_ARGS(&renderPipelineState)));
```
Comment (austinEng) [inline src/backend/d3d12/PipelineD3D12.cpp]: This probably should not have been replaced :P

File: src/backend/d3d12/GeneratedCodeIncludes.h [hunk 4]
```diff
@@ -17,11 +17,12 @@
...
 #include "backend/d3d12/FramebufferD3D12.h"
 #include "backend/d3d12/InputStateD3D12.h"
-#include "backend/d3d12/PipelineD3D12.h"
 #include "backend/d3d12/PipelineLayoutD3D12.h"
 #include "backend/d3d12/QueueD3D12.h"
+#include "backend/d3d12/RenderPipelineD3D12.h"
 #include "backend/d3d12/SamplerD3D12.h"
 #include "backend/d3d12/ShaderModuleD3D12.h"
-#include "backend/d3d12/TextureD3D12.h"
+#include "backend/d3d12/TextureD3D12.h"
```
Comment (austinEng) [inline src/backend/d3d12/GeneratedCodeIncludes.h]: nit: newline at EOF

File: src/backend/d3d12/CommandBufferD3D12.cpp [hunk 5]
```diff
@@ -52,12 +53,17 @@ namespace d3d12 {
             DescriptorHeapHandle cbvSrvUavGPUDescriptorHeap;
             DescriptorHeapHandle samplerGPUDescriptorHeap;
             std::array<BindGroup*, kMaxBindGroups> bindGroups = {};
+            bool inCompute;
```
Comment (austinEng) [inline src/backend/d3d12/CommandBufferD3D12.cpp]: `bool inCompute = false`;

or perhaps `bindingTracker.SetInComputePass(false);` in `Command::BeginRenderPass`

File: src/backend/metal/CommandBufferMTL.mm (hunk 1 above)
Comment (Kangz) [inline src/backend/metal/CommandBufferMTL.mm]: Done

File: src/backend/RenderPipeline.cpp (hunk 2 above)
Comment (Kangz) [inline src/backend/RenderPipeline.cpp]: Done

File: src/backend/d3d12/CommandBufferD3D12.cpp (hunk 5 above)
Comment (Kangz) [inline src/backend/d3d12/CommandBufferD3D12.cpp]: Good catch ! Done.

File: src/backend/d3d12/GeneratedCodeIncludes.h (hunk 4 above)
Comment (Kangz) [inline src/backend/d3d12/GeneratedCodeIncludes.h]: Done.

File: src/backend/d3d12/PipelineD3D12.cpp (hunk 3 above)
Comment (Kangz) [inline src/backend/d3d12/PipelineD3D12.cpp]: Done

**Pull request merged**